packages/stratimikos/adapters/belos/tpetra/Thyra_BelosTpetraPreconditionerFactory_def.hpp

```diff
@@ -0,0 +1,369 @@
+#endif
+
+  // Retrieve concrete preconditioner object
+
+  const Teuchos::Ptr<DefaultPreconditioner<scalar_type> > defaultPrec =
+    Teuchos::ptr(dynamic_cast<DefaultPreconditioner<scalar_type> *>(prec));
+  TEUCHOS_TEST_FOR_EXCEPT(Teuchos::is_null(defaultPrec));
+
+  // This check needed to address Issue #535. 
+  Teuchos::RCP<const Teuchos::ParameterList> innerParamList;
+  if (paramList_.is_null ()) {
+    innerParamList = Teuchos::rcp(new Teuchos::ParameterList(*getValidParameters()));
+  }
+  else {
+    innerParamList = paramList_;
+  }
+
+  bool useHalfPrecision = false;
+  if (innerParamList->isParameter("half precision"))
+    useHalfPrecision = innerParamList->get<bool>("half precision");
+
+  const std::string solverType = Teuchos::getParameter<std::string>(*innerParamList, "BelosPrec Solver Type");
+  const Teuchos::RCP<Teuchos::ParameterList> packageParamList = 
+          Teuchos::rcp(new Teuchos::ParameterList(*Teuchos::sublist(innerParamList, "BelosPrec Solver Params")));
+
+  // solverTypeUpper is the upper-case version of solverType.
+  std::string solverTypeUpper (solverType);
+  if (solverTypeUpper.size () > 0) {
+    for (size_t k = 0; k < solverTypeUpper.size (); ++k) {
+      solverTypeUpper[k] = ::toupper(solverTypeUpper[k]);
+    }
+  }
+
+  // Create the initial preconditioner
+  if (Teuchos::nonnull(out) && Teuchos::includesVerbLevel(verbLevel, Teuchos::VERB_LOW)) {
+    *out << "\nCreating a new BelosTpetra::Preconditioner object...\n";
+  }
+  Teuchos::RCP<LinearOpBase<scalar_type> > thyraPrecOp;
+
+  if (useHalfPrecision) {
+#ifdef THYRA_BELOS_PREC_ENABLE_HALF_PRECISION
+    if (Teuchos::nonnull(out) && Teuchos::includesVerbLevel(verbLevel, Teuchos::VERB_LOW)) {
+      Teuchos::OSTab(out).o() << "> Creating half precision preconditioner\n";
+    }
+    const Teuchos::RCP<const MatrixType> tpetraFwdMatrix = Teuchos::rcp_dynamic_cast<const MatrixType>(tpetraFwdOp);
+    TEUCHOS_TEST_FOR_EXCEPT(Teuchos::is_null(tpetraFwdMatrix));
+    auto tpetraFwdMatrixHalf = tpetraFwdMatrix->template convert<half_scalar_type>();
+
+    Teuchos::RCP<BelosTpLinProbHalf> belosLinProbHalf = Teuchos::RCP<BelosTpLinProbHalf>(new BelosTpLinProbHalf());
+    belosLinProbHalf->setOperator(tpetraFwdMatrixHalf);
+    Teuchos::RCP<TpetraLinOpHalf> belosOpRCPHalf = Teuchos::RCP<TpetraLinOpHalf>(new BelosTpOpHalf(belosLinProbHalf, packageParamList, solverType, true));
+    RCP<TpetraLinOp> wrappedOp = rcp(new Tpetra::MixedScalarMultiplyOp<scalar_type,half_scalar_type,local_ordinal_type,global_ordinal_type,node_type>(belosOpRCPHalf));
+
+    thyraPrecOp = Thyra::createLinearOp(wrappedOp);
+#else
+    if (Teuchos::nonnull(out)) {
+      *out << "\nSolver does not have correct precisions enabled to use half precision.\n";
+    }
+    TEUCHOS_TEST_FOR_EXCEPT(true);
+#endif
+  } else {
+    // Wrap concrete preconditioner
+    Teuchos::RCP<BelosTpLinProb> belosLinProb = Teuchos::RCP<BelosTpLinProb>(new BelosTpLinProb());
```

Comment:
Same here.  Why not:

```
const Teuchos::RCP<BelosTpLinProb> belosLinProb = rcp(new BelosTpLinProb());
```

?

Or, better yet, a non-member constructor function like:

```
const Teuchos::RCP<BelosTpLinProb> belosLinProb = BelosTpLinProb::create();
```

Would be nice to eliminate raw keyword `new` as much as possible.